app/controllers/application_controller/filter.rb

```diff
@@ -0,0 +1,156 @@
+class ApplicationController
+  class Filter
+    INTEGER_COMPARISON_KEYWORDS = ["eq", "gt", "gte", "lt", "lte", "nil", "not_nil"].freeze
+    STRING_COMPARISON_KEYWORDS  = ["contains", "eq", "starts_with", "ends_with", "nil", "not_nil"].freeze
+
+    attr_reader :apply, :arel_table, :api_doc_definition
+
+    def initialize(model, raw_filter, api_doc_definition)
+      self.query          = model
+      @arel_table         = model.arel_table
+      @raw_filter         = raw_filter
+      @api_doc_definition = api_doc_definition
+    end
+
+    def apply
+      return query if @raw_filter.blank?
+      @raw_filter.each do |k, v|
+        next unless attribute = attribute_for_key(k)
+
+        if attribute["type"] == "string"
+          type = determine_string_attribute_type(attribute)
+          send(type, k, v)
+        else
+          errors.add("unsupported attribute type for: #{k}")
+        end
+      end
+
+      raise ApplicationController::Filter::Error.new(:error_document => errors) unless errors.blank?
+      query
+    end
+
+    private
+
+    attr_accessor :query
+
+    class Error < ArgumentError
+      attr_reader :error_document
+
+      def initialize(attrs)
+        @error_document = attrs[:error_document]
+      end
+    end
+
+    def attribute_for_key(key)
+      attribute = api_doc_definition.properties[key.to_s]
+      return attribute if attribute
+      errors.add("found unpermitted parameter: #{key}")
+      nil
+    end
+
+    def determine_string_attribute_type(attribute)
+      return :timestamp if attribute["format"] == "date-time"
+      return :integer if attribute["pattern"] == /^\d+$/
+      :string
+    end
+
+    def errors
+      @errors ||= TopologicalInventory::Api::ErrorDocument.new
+    end
+
+    def string(k, val)
+      if val.kind_of?(ActionController::Parameters)
+        val.each do |comparator, value|
+          add_filter(comparator, STRING_COMPARISON_KEYWORDS, k, value)
+        end
+      else
+        add_filter("eq", STRING_COMPARISON_KEYWORDS, k, val)
+      end
+    end
+
+    def add_filter(requested_comparator, allowed_comparators, key, value)
+      return unless attribute = attribute_for_key(key)
+      type = determine_string_attribute_type(attribute)
+
+      if requested_comparator.in?(["nil", "not_nil"])
```

Comment:
super minor, but we should extract these into a NIL_COMPARISON_KEYWORDS constants in a follow up PR.